File: src/circularsinglylinklist.cpp
#include <iostream>
#include <cassert>
#include <climits>

#include "../include/circularsinglylinklist.h"

// Default constructor
CircularSinglyLinkList::CircularSinglyLinkList()
    : head {NULL}, tail {NULL}, count {0} {
}

// Evaluate if the singly link list has any element
bool CircularSinglyLinkList::isEmpty() const {
    return head == NULL;
}

// Desctructor
CircularSinglyLinkList::~CircularSinglyLinkList() {
    clear();
}

// Remove all nodes in the singly link list
void CircularSinglyLinkList::clear() {

    tail->next = NULL;

    for(Node *ptr; !isEmpty();) {
        ptr = head->next;
        delete head;
        head = ptr;
    }
    count = 0;
}

// Get the number of list elements
int CircularSinglyLinkList::getCount() {
    return count;
}

// Get the value of the head element
int CircularSinglyLinkList::getHeadData() {
    return head->value;
}

// Get the value of the tail element
int CircularSinglyLinkList::getTailData() {
    return tail->value;
}

// Get head address
Node *CircularSinglyLinkList::getPtrHead() {
    return head;
}

// Get tail address
Node *CircularSinglyLinkList::getPtrTail() {
    return tail;
}

// Verify if the singly link list has elements
void CircularSinglyLinkList::emptyness() {
    assert(!isEmpty());
}

// Print all elements the sibgly link list
void CircularSinglyLinkList::display() {

    using std::cout;

    // If there are not elements
    if (isEmpty()) {
        cout << "The circular singly linked list is empty.\n";
        return;
    }

    Node *ptr {head};
    cout << "Elements in the circular singly linked list: \n";

    for (int i {0}; i < count; ++i) {
        cout << ptr->value << " ";
        ptr = ptr->next;
    }

    cout << "\n";
}

void CircularSinglyLinkList::verifyIndex(int index) {
    assert(index >= 0 && index < count);
}

// * Adds new elements

// Adds a new element at beginning of singly link list
void CircularSinglyLinkList::push(int value) {

    // Create a new node and references to its the head as next element
    head = new Node(value, head);

    // When is empty the tail is the new element
    if (tail == NULL)
        tail = head;
    else
        tail->next = head;

    ++count;
}

// Adds a new element at end of singly link list
void CircularSinglyLinkList::append(int value) {

    // If list is not empty
    if (tail != NULL) {
        tail->next = new Node(value);
        tail = tail->next;
        tail->next = head;
    } else
        head = tail = new Node(value);

    ++count;
}

// Adds a new element at a given index in the singly link list
void CircularSinglyLinkList::insert(int index, int value) {

    // Verify if the singly link list has elements
    emptyness();

    // Verification for no more elements in the singly link list
    verifyIndex(index);

    // Add new element
    if (index == 0)
        push(value);
    else if (index == count-1)
        append(value);
    else {

        // Iterate through the desired index
        Node *ptr {head}, *temp {nullptr};

        while (index > 0) {
            temp = ptr;
            ptr = ptr->next;
            --index;
        }
        // Add a new element
        temp->next = new Node(value, ptr);
        ++count;
    }
}

// Adds a new element in a sorted singly link list
void CircularSinglyLinkList::insertSorted(int value) {

    // Existance of elements in the singly link list
    emptyness();

    // Check if the new value is less than head
    if (head->value >= value)
        push(value);
    else if (tail->value <= value)
        append(value);
    else {

        Node *ptr {head}, *temp {head->next};

        while (temp != tail && ptr->value <= value && !(temp->value > value)) {
            ptr = temp;
            temp = temp->next;
        }

        ptr->next = new Node(value, temp);
        ++count;
    }
}

// Adds a new element before the first apperance of the first ocurrance of a given element
void CircularSinglyLinkList::beforeValue(int previousValue, int value) {

    emptyness();

    // Check the first element
    if (head->value == previousValue)
        push(value);
    else {

        Node *ptr {head}, *temp {nullptr};

        while (temp != tail && ptr->value != previousValue) {
            temp = ptr;
            ptr = ptr->next;
        }

        if (ptr != head) {
            temp->next = new Node(value, ptr);
            ++count;
        }
    }
}

// Add a new element after the first apperance of a given value
void CircularSinglyLinkList::afterValue(int afterValue, int value) {

    emptyness();

    // There is only one element
    if (count == 1 && head->value == afterValue)
        append(value);
    else {

        Node *ptr {head}, *temp {head->next};

        while (temp != tail && ptr->value != afterValue) {
            ptr = temp;
            temp = temp->next;
        }

        if (temp == tail && temp->value <= afterValue)
            append(value);
        else if (ptr->value <= afterValue) {
            ptr->next = new Node(value, temp);
            ++count;
        }
    }
}

// * Search methods

// Apply the linear search to the singly link list
Node *CircularSinglyLinkList::search(int value) {

    emptyness();

    // Check the extremes
    if (head->value == value)
        return head;
    else if (tail->value == value)
        return tail;

    Node *ptr {head}, *temp {nullptr};

    for (int i {0}; i < count; ++i) {

        if (ptr->value == value) {
            temp->next = ptr->next;
            ptr->next = head;
            head = ptr;
            tail->next = head;

            return ptr;
        }

        temp = ptr;
        ptr = ptr->next;
    }

    // In case its not found
    return NULL;
}

// * Methods of deletion

int CircularSinglyLinkList::shift() {

    // Check if the singly link list has elements
    emptyness();

    Node *ptr {head};
    int x {ptr->value};

    if (count == 1) {
        clear();
        return x;
    }

    head = head->next;
    tail->next = head;
    delete ptr;
    --count;

    return x;
}

// Removes the last element in the list
int CircularSinglyLinkList::pop() {

    emptyness();

    // If there is just one element
    if (count == 1)
        return shift();

    Node *ptr {head}, *temp {head->next};

    while (temp != tail) {
        ptr = temp;
        temp = temp->next;
    }

    int x {temp->value};
    ptr->next = head;
    tail = ptr;
    delete temp;
    --count;

    return x;
}

// Delete an element by a given index
int CircularSinglyLinkList::Delete(int index) {

    emptyness();
    verifyIndex(index);

    // Verify the extremes
    if (index == 0)
        return shift();
    else if (index == count-1)
        return pop();
    else {

        Node *ptr {head}, *temp {nullptr};

        for (int i {0}; i < index; ++i) {
            temp = ptr;
            ptr = ptr->next;
        }

        int x {ptr->value};
        temp->next = ptr->next;
        delete ptr;
        --count;

        return x;
    }
}

// Remove duplicates
void CircularSinglyLinkList::removeDuplicatesSorted() {

    emptyness();

    Node *ptr {head}, *temp {head->next};

    while (temp !=tail) {

        if (ptr->value != temp->value) {
            ptr = temp;
            temp = temp->next;
        } else {
            ptr->next = temp->next;
            delete temp;
            --count;
            temp = ptr->next;
        }
    }

    if (temp == tail && ptr->value == temp->value) {
        tail = ptr;
        ptr->next = head;
        delete temp;
        --count;
    }
}

// Delete the first apperance of the given value
int CircularSinglyLinkList::remove(int value) {

    emptyness();

    // In case its the first element
    if (head->value == value)
        return shift();

    Node *ptr {head}, *temp {nullptr};
    int x {INT_MIN};

    while (ptr != tail && ptr->value != value) {
        temp = ptr;
        ptr = ptr->next;
    }

    // Verify the last pair of elements
    if (ptr == tail && ptr->value == value) {
        x = ptr->value;
        tail = temp;
        temp->next = head;
        delete ptr;
        --count;
        return x;
    // In case it is found before the last two elements
    } else if (ptr->value == value && ptr != tail) {
        x = ptr->value;
        temp->next = ptr->next;
        delete ptr;
        --count;
        return x;
    }

    return x;
}

// * Methods of sorting

// Check if the singly link list is sorted
bool CircularSinglyLinkList::checkSort() {

    emptyness();

    Node *ptr {head};
    int x {INT_MIN};

    for (int i {0}; i < count; ++i) {
        if (ptr->value < x)
            return false;

        x = ptr->value;
        ptr = ptr->next;
    }

    return true;
}

void CircularSinglyLinkList::reverse() {

    emptyness();

    Node *ptr {head}, *prevNode {nullptr}, *nextNode {nullptr};
    tail = head;

    for (int i {0}; i < count; ++i) {
        nextNode = ptr->next;
        ptr->next = prevNode;
        prevNode = ptr;
        ptr = nextNode;
    }

    head = prevNode;
    tail = ptr;
    ptr->next = head;
}

// Check if the singly link list is looped otherwise is linear
bool CircularSinglyLinkList::isLoop() {

    emptyness();

    Node *ptr {head}, *temp {head};

    do {
        ptr = ptr->next;
        temp = temp->next;
        temp = (temp) ? temp->next : NULL;
    } while (ptr && temp && ptr != temp);

    return (ptr == temp) ? true : false;
}

// * Methods of update

// Replace all values or the first apperance of a given value with a new one
void CircularSinglyLinkList::replace(int oldValue, int value, bool entire) {

    emptyness();

    Node *ptr {head};

    for (int i {0}; i < count; ++i) {
        // Replace the value
        if (ptr->value == oldValue) {
            ptr->value = value;

            // Stop in case the entire value is false
            if (!entire)
                break;
        }

        ptr = ptr->next;
    }
}

// Replace an value with a new value by a given index
void CircularSinglyLinkList::update(int index, int value) {

    // Existance of the elements in the list
    emptyness();
    // Check the range of the index
    verifyIndex(index);

    Node *ptr {head};

    for (int i {0}; i < index; ++i)
        ptr = ptr->next;

    ptr->value = value;
}

File: python/circularsinglylinklist.py
from sys import maxsize
from dataclasses import dataclass
from node import Node
from typing import Optional


@dataclass
class CircularSingleLinkedList:
    pass


    def __post_init__(self):
        self.__head: Node = None
        self.__tail: Node = None
        self.__count: int = 0


    @property
    def head(self) -> Node:
        """Gets the property of the head of the circular singly linked list."""
        return self.__head


    @property
    def tail(self) -> Node:
        """Gets the property of the tail of the circular singly linked list."""
        return self.__tail


    @property
    def count(self) -> int:
        """Gets the number of items in the list"""
        return self.__count


    def __len__(self) -> int:
        """Returns the count of nodes in the circular singly linked list."""
        return self.__count


    def __str__(self) -> str:
        """Returns the string representation of the circular singly linked list."""
        return f'CircularSinglyLinkedList(head={self.__head}, tail={self.__tail}, count={self.__count})'


    def __existance_of_elements(self):
        """Asserts if there is any node in the list."""
        ## Evaluate the existence of the single linked list
        assert self.head, 'There are no elements in the circular singly linked list. Use the add or append method.'


    def __iter__(self):
        """Iterates through the all nodes of the circular singly linked list."""
        ## If there is not head
        self.__existance_of_elements()

        ## Iterate through all nodes of the circular single linked list
        probe: Node = self.head

        for _ in range(self.count):
            yield probe
            probe = probe.next


    def display(self):
        """Display all nodes values of the circular singly linked list."""
        for node in self.__iter__():
            print(node.value)


    def __verify_index(self, index: int):
        """Verifies the index of the input."""
        if not(-1 <= index < self.count):
            raise IndexError(f'The index value: {index} is out of range in the circular singly linked list.')


    def push(self, value: int):
        """Adds a new node to the begging of the circular singly linked list."""

        self.__head = Node(value, self.__head)

        if not self.tail:
            self.__tail = self.head
        else:
            self.__tail.next = self.head

        self.__count += 1


    def append(self, value: int):
        """Appends new node to the end of the circular singly linked list."""

        if self.tail:
            self.__tail.next = Node(value)
            self.__tail = self.__tail.next
            self.__tail.next = self.head
        else:
            self.__head = self.__tail = Node(value)

        self.__count += 1


    def insert(self, index: int, value: int):
        """Inserts a new node in function of the index."""
        ## Check if the list exist
        self.__existance_of_elements()
        ## Verify if list is empty
        self.__verify_index(index)

        ## Inserts on the index
        if index == 0:
            ## Add the beginning
            self.push(value)
        elif index in [-1, self.count-1]:
            ## Add the end of the single linked list
            self.append(value)
        else:
            probe: Node = self.head
            temp: Node = None

            for _ in range(index):
                temp = probe
                probe = probe.next

            temp.next = Node(value, probe)
            self.__count += 1


    def before_value(self, before_value: int, value: int):
        """Inserts a new node before the first appearance of a given value."""
        ## Evaluate if the single linked list is empty
        self.__existance_of_elements()

        ## Evaluate the insertion
        if self.head.value == before_value:
            self.push(value)
        else:
            probe: Node = self.head
            temp: Node = None

            while temp != self.tail and probe.value != before_value:
                temp = probe
                probe = probe.next

            if probe != self.head:
                temp.next = Node(value, probe)
                self.__count += 1


    def after_value(self, after_value: int, value: int):
        """Inserts a new node after the first appearance of given value."""
        ## Evaluate the existance
        self.__existance_of_elements()

        if self.count == 1 and self.head.value == after_value:
            self.append(value)
        else:
            probe: Node = self.head
            temp: Node = self.head.next

            while temp != self.tail and probe.value != after_value:
                probe = temp
                temp = temp.next

            if temp == self.tail and temp.value <= after_value:
                self.append(value)
            elif probe.value <= after_value:
                probe.next = Node(value, temp)
                self.__count += 1


    def insert_sorted(self, value: int):
        """Insert a new element in a sorted in the circular singly linked list."""
        self.__existance_of_elements()

        if self.head.value >= value:
            self.push(value)
        elif self.tail.value <= value:
            self.append(value)
        else:
            probe: Node = self.head
            temp: Node = self.head.next

            while temp != self.tail and probe.value <= value and not(temp.value > value):
                probe = temp
                temp = temp.next

            probe.next = Node(value, temp)
            self.__count += 1



    def index(self, value: int) -> list[int]:
        """Search for the index of the given value and return the list of indixes"""
        return [
            i for i, node in enumerate(self.__iter__()) if node.value == value
        ]


    def clear(self):
        """Deletes the circular singly linked list."""
        self.__head = None
        self.__tail = None
        self.__count = 0


    def pop(self):
        """Deletes the last node of the circular singly linked list."""
        ## Evaluates if the list is empty
        self.__existance_of_elements()

        ## Evaluates the number of nodes in the list
        if self.count == 1:
            self.clear()
            return

        ## Iterate to the node before the last
        probe: Node = self.head
        temp: Node = self.head.next

        while temp != self.tail:
            probe = temp
            temp = temp.next

        probe.next = self.head
        self.__tail = probe
        self.__count -= 1


    def shift(self):
        """Deletes the first node of the circular singly linked list."""
        ## Evaluates if the list is empty
        self.__existance_of_elements()

        if self.count == 1:
            self.clear()
            return

        self.__head = self.head.next
        self.__tail.next = self.head
        self.__count -= 1


    def delete(self, index: int):
        """Deletes a node by a given index."""
        self.__existance_of_elements()
        ## Verify the index
        self.__verify_index(index)

        ## Verify the index corners
        if index == 0:
            self.shift()
        elif index in [-1, self.count-1]:
            self.pop()
        else:
            probe: Node = self.head
            temp: Node = None

            for _ in range(index):
                temp = probe
                probe = probe.next

            temp.next = probe.next
            self.__count -= 1


    def remove(self, value: int):
        """Deletes the first appearance of the given value."""
        ## Evaluates if the list is empty
        self.__existance_of_elements()

        ## In case it's the first element
        if self.head.value == value:
            self.shift()
            return

        probe: Node = self.head
        temp: Node = None

        while probe != self.tail and probe.value != value:
            temp = probe
            probe = probe.next

        if probe == self.tail and probe.value == value:
            self.__tail = temp
            temp.next = self.head
            self.__count -= 1
        elif probe.value == value and probe != self.tail:
            temp.next = probe.next
            self.__count -= 1


    def unique_sorted(self):
        """Remove the duplicate values from the circular singly linked list."""
        ## Check the existence of the list
        self.__existance_of_elements()

        probe: Node = self.head
        temp: Node = self.head.next

        while temp != self.tail:
            if probe.value != temp.value:
                probe = temp
                temp = temp.next
            else:
                probe.next = temp.next
                self.__count -= 1
                temp = probe.next

        if temp == self.tail and probe.value == temp.value:
            self.__tail = probe
            probe.next = self.head
            self.__count -= 1


    def replace(self, old_value: int,value: int, entire: bool = False):
        """Replaces the first occurrence of the given value with new one or all apperances. By default replace the first appearance."""
        ## Evaluates if the list is empty
        self.__existance_of_elements()

        ## Iterate through all the elements in case the all entire parameter is true
        for node in self.__iter__():
            if node.value == old_value:
                node.value = value
                if not entire:
                    break


    def update(self, index: int, value: int):
        """Updates a value in the list by a given index."""
        self.__existance_of_elements()
        ## Evaluates the correct index
        self.__verify_index(index)

        probe: Node = self.head

        for _ in range(index):
            probe = probe.next
        
        ## Updates the value
        probe.value = value


    def reverse(self):
        """Reverse the circular singly linked list backwards."""
        ## Evaluates if the list is empty
        self.__existance_of_elements()

        next_node: Node = None
        prev_node: Node = None
        probe: Node = self.head

        for _ in range(self.count):
            next_node = probe.next ## Keep the next node of the original list
            probe.next = prev_node ## Join the next property with the previous node
            prev_node = probe ## Move the previous as the current
            probe = next_node ## Move to the next node of the original list

        self.__head = prev_node
        self.__tail = probe
        probe.next = self.head


    def check_sort(self) -> bool:
        """Check if the circular singly linked list is sorted."""
        self.__existance_of_elements()

        probe: Node = self.head
        x: int = -maxsize - 1

        for _ in range(self.count):
            if probe.value < x:
                return False

            x = probe.value
            probe = probe.next

        return True


    def search(self, value: int) -> Optional[Node]:
        """Returns the a node if the value is in the list."""
        self.__existance_of_elements()

        ## Check the extremes
        if self.head.value == value:
            return self.head
        elif self.tail.value == value:
            return self.tail

        probe: Node = self.head
        temp: Node = None

        for _ in range(self.count):

            if probe.value == value:
                temp.next = probe.next
                probe.next = self.head
                self.__head = probe
                self.__tail.next = self.head
                return probe

            temp = probe
            probe = probe.next

        return None